pkg/apis/apps/v1alpha1/uniteddeployment_types.go

```diff
@@ -0,0 +1,229 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// UnitedDeploymentSpec defines the desired state of UnitedDeployment
+type UnitedDeploymentSpec struct {
+	// Replicas is the totally desired number of replicas of all the owning workloads.
+	// If unspecified, defaults to 0.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// selector is a label query over pods that should match the replica count.
+	// It must match the pod template's labels.
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// Template is the object that describes the subset that will be created
+	Template SubsetTemplate `json:"template,omitempty"`
+
+	// Contains the information of subset topology
+	Topology Topology `json:"topology,omitempty"`
+
+	Strategy UnitedDeploymentUpdateStrategy `json:"strategy,omitempty"`
+
+	// Indicate the number of histories to be conserved
+	// If unspecified, defaults to 20
+	// +optional
+	RevisionHistoryLimit *int32 `json:"historyLimit,omitempty"`
+}
+
+// SubsetTemplate defines the subset under the UnitedDeployment
+type SubsetTemplate struct {
+	// Indicate the type of the underlying workload
+	// If unspecified, defaults to ReplicaSet
+	// +optional
+	SubsetType *metav1.TypeMeta `json:"subsetType,omitempty"`
+
+	// StatefulSet template
+	StatefulSet *StatefulSetTemplateSpec `json:"statefulSet,omitempty"`
+
+	// ReplicaSet template
+	ReplicaSet *ReplicaSetTemplateSpec `json:"replicaSet,omitempty"`
+}
+
+// StatefulSetTemplateSpec defines the subset template of StatefulSet
+type StatefulSetTemplateSpec struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              appsv1.StatefulSetSpec `json:"spec,omitempty"`
+}
+
+// ReplicaSetTemplateSpec defines the subset template of ReplicasSet
+type ReplicaSetTemplateSpec struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              appsv1.ReplicaSetSpec `json:"spec,omitempty"`
+}
+
+// UnitedDeploymentUpdateStrategy defines the update strategy of UnitedDeployment
+type UnitedDeploymentUpdateStrategy struct {
+	// Pause the rollout of all subset worloads
+	// +optional
+	Pause bool `json:"pause,omitempty"`
+
+	// Indicate the partition of each subset
+	Partitions map[string]int32 `json:"partitions,omitempty"`
+}
+
+// Topology defines the spread detail of each subset under UnitedDeployment
+type Topology struct {
+	// Contains the details of each subset
+	Subsets []Subset `json:"subsets,omitempty"`
+}
+
+// Subset defines the detail of a subset
+type Subset struct {
+	// Indicate the name of this subset, which will be used to generate
+	// subset workload name in the format '<deployment-name>-<subset-name>'
+	Name string `json:"name,omitempty"`
```

Comment:
If this is unset, what is the default subset name?

Reply:
I suggest it to be required.